API/models/Posts.js

```diff
@@ -0,0 +1,177 @@
+const mongoose = require('mongoose');
+const Schema = mongoose.Schema;
+
+// TODO: implement comments model
+// require('./Comments');
+// const Comment = mongoose.model('Comment');
+
+const PostSchema = new Schema({
+	author: {
+    //type: Schema.Types.ObjectId,
+    type: String,
+    required: true,
+    ref: 'User'
+  },
+  subscribedUsers: {
+    type: Array,
+    default: [],
+    ref: 'User'
+  },
+	content: {
+		type: String,
+		required: true,
+  },
+  image: {
+    type: Buffer,
+  },
+	tags: {
+		type: Array,
+		required: true,
+  },
+  comments: {
+    type: Array,
+    ref: 'Comment'
+  },
+  createdAt: {
+    type: Object,
+  },
+  updatedAt: {
+    type: Object,
+  }
+});
+
+PostSchema.statics.create = function(post) {
+  // Implement a character limit client-side
+  if (!post.createdAt) {
+    post.createdAt = new Date();
+  }
+	var formattedTags = [];
+	post.tags.forEach((tag) => {
+    formattedTags.push(tag.toLowerCase());
+  });
+  var post = {
+    ...post,
+    subscribedUsers: post.author,
+    tags: formattedTags,
+  }
+	return new Promise((resolve, reject) => {
+    this.find({ content: post.content }).then(existingPost => {
```

Comment:
It seems a little odd to me to be searching by a long string of text, though I could very well be missing something.

Reply:
Right, my thought here was avoiding duplicate posts in case of, for example, the text input area on the frontend not clearing on submit (prompting the user to click submit again). The problem with this scenario is the timestamp for the duplicate posts would be different, so the new document object would not be a true duplicate of the previous one. It is unlikely the user would change the content of the post in such a scenario (more likely just spam the submit button), so querying by post content would likely avoid duplication.

With that said, it is a bit strange to query by a large string, so there might be a better way of doing this. LMK if you have any ideas :) (for example could do 'debounce' logic to prevent the user from posting twice within, say, a minute - seems unnecessarily restrictive though).

Reply:
(It might also be better convention to 'trust' posts to the database and handle this kind of scenario client-side)

Reply:
Ah, I see. I think the best way to check post uniqueness would be to calculate the hash value of the string. NoSQL might have a method of handling that for us the way MySQL does using indexes, we'll need to look into that. That said, there might be a case where we don't want to enforce uniqueness, e.g. I could see "Veebs at 5:30. Be there." being intentionally posted twice on different days. In cases like that it might make more sense to not enforce uniqueness of posts, but merely let users delete their own posts.

In general I'd say never trust anything that comes from the client side. It's very easy for an attacker or spammer to get around any client-side validations and just make their own calls to our api with bad data. We can do things like disallowing empty posts client side, but we also need to do them server side in case someone gets around the client side checks (I don't really think anyone will care enough to try and hack us, but it never hurts to follow good conventions). There are also the benign cases where we forget to validate something and then people never update the app to the version that has the validation, in which case we need server side validations anyways.

Reply:
Personally I don't think duplicate posts is a concern of ours here. From the API point of view If people want to spam the API with posts thats fine.

If our app however accidentally posts multiple posts that's a problem with the app - not with the API so should be addressed there.

Reply:
We could limit how frequently a user can post (ex. must wait 10 seconds before posting again or something) but I don't think its worth worrying about